### PR TITLE
[rust] Support different ways to configure Selenium Manager (config file, envs)

### PR DESCRIPTION
**Pull request by bonigarcia**
### Description
This PR implements #11690 and #11689. I know the latter was closed because it was not unwanted at this point, but for a similar effort, and because I honestly think it might help some users out there, I have implemented both.

So far, the Selenium Manager configuration capabilities have been done through CLI arguments. For instance, the flag `--proxy <PROXY>` allows configuring the proxy for the network connection. Nevertheless, this has the inconvenience that to make this configuration possible, all the bindings languages should implement a mechanism (e.g., through browser options) to make this feature available for users. For instance, the configuration before (i.e., the proxy) is not possible in Selenium Manager at the time of this writing since it has yet to be implemented in the bindings.

This PR enhances the configuration capabilities by providing the procedure proposed in #11690:

1. CLI parameters. 
2. Configuration file.
3. Environment variables.

IMO, option two can be interesting for global configuration in a local machine (e.g., a developer laptop). Option three can be interesting for global configuration in build servers (e.g., Jenkins, GitHub Actions, etc.). Both of them allow a rapid configuration (since it bypasses the need to implement it in the bindings).

Here are some examples of configuration:

a) For specifying a fixed version of geckodriver:
a.1) With config file:
```
boni@slimbook:~/.cache/selenium$ cat selenium-manager-config.toml 
geckodriver-version = "0.30.0"


boni@slimbook:~/dev/selenium/rust$ cargo run -- --browser firefox --debug
    Finished dev [unoptimized + debuginfo] target(s) in 0.08s
     Running `target/debug/selenium-manager --browser firefox --debug`
DEBUG	Running command: "geckodriver --version"
DEBUG	Output: ""
DEBUG	Driver URL: https://github.com/mozilla/geckodriver/releases/download/v0.30.0/geckodriver-v0.30.0-linux64.tar.gz
INFO	/home/boni/.cache/selenium/geckodriver/linux64/0.30.0/geckodriver
```

a.2) With envs:
```
boni@slimbook:~/dev/selenium/rust$ SE_GECKODRIVER_VERSION=0.30.0 cargo run -- --browser firefox --debug
    Finished dev [unoptimized + debuginfo] target(s) in 0.08s
     Running `target/debug/selenium-manager --browser firefox --debug`
DEBUG	Running command: "geckodriver --version"
DEBUG	Output: ""
DEBUG	Driver URL: https://github.com/mozilla/geckodriver/releases/download/v0.30.0/geckodriver-v0.30.0-linux64.tar.gz
INFO	/home/boni/.cache/selenium/geckodriver/linux64/0.30.0/geckodriver
```

b) For specifying a proxy:

b.1) With config file:
```
boni@slimbook:~/.cache/selenium$ cat selenium-manager-config.toml 
proxy = "myproxy:8080"
```

b.2) With envs:
```
SE_PROXY=myproxy:8080 ./selenium-manager --browser chrome
```

### Motivation and Context
This PR is part of M3 (rich configuration) in the Selenium Manager roadmap.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [X] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by bonigarcia** [comment]
Just a note to clarify the naming of the configuration keys.

In the configuration file (`~/.cache/selenium/selenium-manager-config.toml`), each configuration key has the same name as the CLI arguments, such as `browser-version`, `driver-version`, `browser-path`, `driver-ttl`, `browser-ttl`, `proxy`, etc. Moreover, there are specific configuration keys for browser versions, driver versions, and browser paths, such as:

- `chrome-version`, `chromedriver-version`, and `chrome-path`
- `firefox-version`, `geckodriver-version`, and `firefox-path`
- `edge-version`, `msedgedriver-version`, and `edge-path`

All of these configuration keys have their equivalence in environmental variables by converting each parameter name to uppercase, replacing the dash symbol (`-`) with an underscore (`_`), and adding the prefix`SE_`. Some examples:

`SE_DRIVER_TTL`, `SE_GECKODRIVER_VERSION`, `SE_FIREFOX_PATH`, `SE_PROXY`, `SE_DEBUG`, etc.

File: rust/Cargo.toml [hunk 1]
```diff
@@ -28,6 +28,7 @@ tar = "0.4.38"
 infer = "0.13.0"
 exitcode = "1.1.2"
 is_executable = "1.0.1"
+toml = "0.7.3"
```
Comment (diemol) [inline rust/Cargo.toml:31]: Does the binary grow too much with this new dependency?

File: rust/src/chrome.rs [hunk 2]
```diff
@@ -49,14 +49,19 @@ pub struct ChromeManager {
...
-            browser_name: CHROME_NAME,
-            driver_name: CHROMEDRIVER_NAME,
-            config: ManagerConfig::default(),
-            http_client: create_default_http_client(),
+    pub fn new() -> Result<Box<Self>, String> {
+        let browser_name = CHROME_NAME;
+        let driver_name = CHROMEDRIVER_NAME;
+        let config = ManagerConfig::default(browser_name, driver_name);
+        let default_timeout = config.timeout.to_owned();
+        let default_proxy = config.proxy.to_owned();
```
Comment (diemol) [inline rust/src/chrome.rs:57]: Just checking, are these things related to the PR description? Should they be part of a different PR?

File: rust/src/config.rs [hunk 3]
```diff
@@ -142,3 +158,74 @@ impl ARCH {
...
+    }
+}
+
+fn get_env_name(suffix: &str) -> String {
+    let suffix_uppercase: String = suffix.replace('-', "_").to_uppercase();
+    concat(ENV_PREFIX, suffix_uppercase.as_str())
+}
+
+fn get_config() -> Result<Table, Box<dyn Error>> {
+    let config_path = get_cache_folder().join(CONFIG_FILE);
```
Comment (diemol) [inline rust/src/config.rs:223]: This means that if the file is present, it will be used, right?

What has priority, env vars or file?

**Review by diemol — COMMENTED** [review]
Left some comments.

File: rust/Cargo.toml (hunk 1 above)
Comment (bonigarcia) [reply rust/Cargo.toml:31]: The crate `toml` is quite small. I have just built the binaries using this PR branch (see [here](https://github.com/SeleniumHQ/selenium/actions/runs/4588075402)), and the resulting binaries are:

selenium-manager_linux-x64	4.31 MB
selenium-manager_macos-x64	3.59 MB
selenium-manager_windows-x64	3.6 MB

The current binaries are (see [here](https://github.com/SeleniumHQ/selenium/actions/runs/4545218225)):

selenium-manager_linux-x64	4.14 MB	
selenium-manager_macos-x64	3.45 MB	
selenium-manager_windows-x64	3.46 MB

... so the difference is around 0.15 MB per binary.

File: rust/src/chrome.rs (hunk 2 above)
Comment (bonigarcia) [reply rust/src/chrome.rs:57]: These changes are required since with this PR, the proxy setup can be different because of the config in the files/envs.

File: rust/src/config.rs (hunk 3 above)
Comment (bonigarcia) [reply rust/src/config.rs:223]: The config file is used when is present and the corresponding CLI parameter is not honored.

In other words, the order of preference is as follows:

1. CLI parameters.
2. Configuration file.
3. Environment variables.

**Comment by bonigarcia** [comment]
I have just rebased this branch with trunk. If this PR is accepted, I recommend to merge it before the others (e.g. #11840), since it is going to create conflicts otherwise.

File: rust/src/config.rs [hunk 4]
```diff
@@ -150,17 +159,24 @@ impl ARCH {
...
 
 impl StringKey<'_> {
     fn get_value(&self) -> String {
         let config = get_config().unwrap_or_default();
-        let key = self.0;
-        if config.contains_key(key) {
-            config[key].as_str().unwrap().to_string()
-        } else {
-            env::var(get_env_name(key)).unwrap_or(self.1.to_owned())
+        let keys = self.0.to_owned();
```
Comment (oriongonza) [inline rust/src/config.rs:167]: You don't need to clone the keys:
```rust
 let keys = &self.0;
 let mut result: String;
 for &key in keys {
```

Also what's the purpose of `.as_str().unwrap().to_string()` instead of `to_string()` directly?

Comment (diemol) [reply rust/src/config.rs:167]: What is the alternative to cloning the keys? Maybe you can suggest the code change?

Comment (oriongonza) [reply rust/src/config.rs:167]: I just did, in the comment block. 
I'm new to contributing to github, I messed something up: I wanted to highlight the block but only highlighted the line. I should have read [the docs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/reviewing-proposed-changes-in-a-pull-request?tool=webui). My bad!

Anyways the change is to just take a reference to the `Vec`,  `let keys = &self.0;` and then dereference the key using pattern matching `for &key in keys`

same effect as 
```rust
        let keys = &self.0;
        for key in keys {
            if config.contains_key(*key) {
                result = config[*key].as_str().unwrap().to_string();
```

File: rust/src/config.rs [hunk 5]
```diff
@@ -150,17 +159,24 @@ impl ARCH {
...
+        let mut result;
+        for key in keys {
+            if config.contains_key(key) {
+                result = config[key].as_str().unwrap().to_string()
```
Comment (oriongonza) [inline rust/src/config.rs:171]:
```suggestion
        let keys = &self.0;
        let mut result;
        for &key in keys {
            if config.contains_key(key) {
                result = config[key].to_string()
```

**Review by oriongonza — CHANGES_REQUESTED** [review]

File: rust/src/config.rs [hunk 6]
```diff
@@ -142,3 +158,74 @@ impl ARCH {
             .contains(&arch.to_ascii_lowercase().as_str())
     }
 }
+
+pub struct StringKey<'a>(pub Vec<&'a str>, pub &'a str);
+
+impl StringKey<'_> {
```
Comment (oriongonza) [inline rust/src/config.rs:164]: Every time StringKey is used it's just used to call get_key() on it. It might make more sense to just convert it to a function eg

```rust
pub struct Keys; 
impl Keys {
  get_string_key(key: &str, default: &str){...}
  }

**Comment by bonigarcia** [comment]
@dev-ardi Can you please send a separate PR for these improvements you mentioned? Thanks a lot!

**Review by diemol — APPROVED** [review]

File: rust/src/lib.rs [hunk 7]
```diff
@@ -508,17 +508,22 @@ pub fn clear_cache(log: &Logger) {
...
         .danger_accept_invalid_certs(true)
         .use_rustls_tls()
+        .timeout(Duration::from_secs(timeout));
+    if !proxy.is_empty() {
+        match Proxy::all(proxy) {
+            Ok(p) => {
+                client_builder = client_builder.proxy(p);
+            }
+            Err(err) => {
+                return Err(err.to_string());
```
Comment (oriongonza) [inline rust/src/lib.rs:522]: @bonigarcia  What was the reason to return a String instead of an Error?

Comment (bonigarcia) [reply rust/src/lib.rs:522]: That's because the final error messages shown as a result are string. But perhaps there is a more idiomatic way to handle this error. Please send a PR if you find a better way to do it.
